File: diagram/ai/_gen_fisik_docx.py
# /// script
# requires-python = ">=3.11"
# dependencies = ["python-docx>=1.1.2"]
# ///
"""Generate 4.3.3 Perancangan Fisik Basis Data .docx aligned to ERD (6 entitas).
Tables: tb_user, tb_klien, tb_proyek, tb_penugasan_proyek, tb_tugas, tb_bukti,
tb_eskalasi. Style: Times New Roman, 5-col schema tables."""
from docx import Document
from docx.shared import Pt, Cm, RGBColor, Twips
from docx.enum.table import WD_ALIGN_VERTICAL, WD_TABLE_ALIGNMENT
from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.oxml import OxmlElement
from docx.oxml.ns import qn

OUT = r"D:\__CODING\05-MyProjects\__IRENE\shi-crm\diagram\ai\4.3.3_PERANCANGAN_FISIK.docx"
FONT = "Times New Roman"

# Column widths in twips (DXA). 5 columns: No | Nama Field | Tipe Data | Panjang | Constraint
COL_W = [600, 2400, 1800, 3360, 1500]

TABLES = [
    {
        "num": "4.6", "title": "User", "nama": "users", "pk": "id", "fk": [],
        "rows": [
            ("id", "SERIAL", "-", "PK"),
            ("name", "VARCHAR", "255", "NOT NULL"),
            ("email", "VARCHAR", "255", "Unique, NOT NULL"),
            ("role", "VARCHAR", "50", "NOT NULL, Default: technician"),
            ("password_hash", "VARCHAR", "255", "NOT NULL"),
            ("created_at", "TIMESTAMP", "-", "Default: CURRENT_TIMESTAMP"),
        ],
    },
    {
        "num": "4.7", "title": "Klien", "nama": "clients", "pk": "id",
        "fk": ["created_by (terhubung dengan users)"],
        "rows": [
            ("id", "SERIAL", "-", "PK"),
            ("name", "VARCHAR", "255", "NOT NULL"),
            ("address", "TEXT", "-", ""),
            ("phone", "VARCHAR", "50", ""),
            ("email", "VARCHAR", "255", ""),
            ("notes", "TEXT", "-", ""),
            ("latitude", "DECIMAL", "10,7", ""),
            ("longitude", "DECIMAL", "10,7", ""),
            ("photo_path", "VARCHAR", "1000", ""),
            ("photo_name", "VARCHAR", "500", ""),
            ("created_by", "INT", "-", "FK, NOT NULL"),
            ("created_at", "TIMESTAMP", "-", "Default: CURRENT_TIMESTAMP"),
            ("updated_at", "TIMESTAMP", "-", "Default: CURRENT_TIMESTAMP"),
        ],
    },
    {
        "num": "4.8", "title": "Proyek", "nama": "projects", "pk": "id",
        "fk": [
            "client_id (terhubung dengan clients)",
            "created_by (terhubung dengan users)",
            "survey_approved_by (terhubung dengan users)",
        ],
        "rows": [
            ("id", "SERIAL", "-", "PK"),
            ("project_code", "VARCHAR", "12", "Unique, NOT NULL"),
            ("name", "VARCHAR", "255", "NOT NULL"),
            ("description", "TEXT", "-", ""),
            ("client_id", "INT", "-", "FK"),
            ("start_date", "DATE", "-", "NOT NULL"),
            ("end_date", "DATE", "-", "NOT NULL"),
            ("duration", "INT", "-", "Generated"),
            ("status", "VARCHAR", "50", "Default: active"),
            ("phase", "VARCHAR", "50", "Default: survey"),
            ("category", "VARCHAR", "50", "Default: instalasi"),
            ("project_value", "DECIMAL", "15,2", "Default: 0"),
            ("survey_approved", "BOOLEAN", "-", "Default: FALSE"),
            ("survey_approved_by", "INT", "-", "FK"),
            ("survey_approved_at", "TIMESTAMP", "-", ""),
            ("target_description", "TEXT", "-", ""),
            ("created_by", "INT", "-", "FK, NOT NULL"),
            ("created_at", "TIMESTAMP", "-", "Default: CURRENT_TIMESTAMP"),
            ("updated_at", "TIMESTAMP", "-", "Default: CURRENT_TIMESTAMP"),
        ],
    },
    {
        "num": "4.9", "title": "Penugasan Proyek", "nama": "project_assignments",
        "pk": "(project_id, user_id)",
        "fk": [
            "project_id (terhubung dengan projects)",
            "user_id (terhubung dengan users)",
        ],
        "rows": [
            ("project_id", "INT", "-", "PK, FK"),
            ("user_id", "INT", "-", "PK, FK"),
            ("assigned_at", "TIMESTAMP", "-", "Default: CURRENT_TIMESTAMP"),
        ],
    },
    {
        "num": "4.10", "title": "Tugas", "nama": "tasks", "pk": "id",
        "fk": [
            "project_id (terhubung dengan projects)",
            "assigned_to (terhubung dengan users)",
            "depends_on (terhubung dengan tasks)",
            "created_by (terhubung dengan users)",
        ],
        "rows": [
            ("id", "SERIAL", "-", "PK"),
            ("project_id", "INT", "-", "FK, NOT NULL"),
            ("name", "VARCHAR", "500", "NOT NULL"),
            ("description", "TEXT", "-", ""),
            ("assigned_to", "INT", "-", "FK"),
            ("status", "VARCHAR", "50", "Default: to_do"),
            ("due_date", "DATE", "-", ""),
            ("timeline_start", "DATE", "-", ""),
            ("timeline_end", "DATE", "-", ""),
            ("notes", "TEXT", "-", ""),
            ("sort_order", "INT", "-", "Default: 0"),
            ("is_survey_task", "BOOLEAN", "-", "Default: FALSE"),
            ("timer_started_at", "TIMESTAMP", "-", ""),
            ("time_spent_seconds", "INT", "-", "Default: 0"),
            ("is_tracking", "BOOLEAN", "-", "Default: FALSE"),
            ("estimated_hours", "DECIMAL", "5,1", ""),
            ("depends_on", "INT", "-", "FK"),
            ("status_changed_at", "TIMESTAMP", "-", "Default: NOW()"),
            ("created_by", "INT", "-", "FK, NOT NULL"),
            ("created_at", "TIMESTAMP", "-", "Default: CURRENT_TIMESTAMP"),
            ("updated_at", "TIMESTAMP", "-", "Default: CURRENT_TIMESTAMP"),
        ],
    },
    {
        "num": "4.11", "title": "Bukti Tugas", "nama": "task_evidence", "pk": "id",
        "fk": [
            "task_id (terhubung dengan tasks)",
            "uploaded_by (terhubung dengan users)",
        ],
        "rows": [
            ("id", "SERIAL", "-", "PK"),
            ("task_id", "INT", "-", "FK, NOT NULL"),
            ("file_path", "VARCHAR", "1000", "NOT NULL"),
            ("file_name", "VARCHAR", "500", "NOT NULL"),
            ("file_type", "VARCHAR", "50", "NOT NULL"),
            ("file_size", "INT", "-", "Default: 0"),
            ("description", "TEXT", "-", ""),
            ("uploaded_by", "INT", "-", "FK, NOT NULL"),
            ("uploaded_at", "TIMESTAMP", "-", "Default: CURRENT_TIMESTAMP"),
        ],
    },
    {
        "num": "4.12", "title": "Daily Report", "nama": "daily_reports", "pk": "id",
        "fk": [
            "project_id (terhubung dengan projects)",
            "task_id (terhubung dengan tasks)",
            "created_by (terhubung dengan users)",
        ],
        "rows": [
            ("id", "SERIAL", "-", "PK"),
            ("project_id", "INT", "-", "FK, NOT NULL"),
            ("task_id", "INT", "-", "FK"),
            ("report_date", "DATE", "-", "NOT NULL"),
            ("progress_percentage", "DECIMAL", "5,2", "NOT NULL, Range: 0-100"),
            ("constraints", "TEXT", "-", ""),
            ("created_by", "INT", "-", "FK, NOT NULL"),
            ("created_at", "TIMESTAMP", "-", "Default: CURRENT_TIMESTAMP"),
        ],
    },
    {
        "num": "4.13", "title": "Kesehatan Proyek", "nama": "project_health",
        "pk": "project_id",
        "fk": ["project_id (terhubung dengan projects)"],
        "rows": [
            ("project_id", "INT", "-", "PK, FK"),
            ("spi_value", "DECIMAL", "6,4", ""),
            ("status", "VARCHAR", "50", "CHECK: green|amber|red"),
            ("actual_progress", "DECIMAL", "5,2", ""),
            ("planned_progress", "DECIMAL", "5,2", ""),
            ("total_tasks", "INT", "-", "Default: 0"),
            ("completed_tasks", "INT", "-", "Default: 0"),
            ("working_tasks", "INT", "-", "Default: 0"),
            ("overtime_tasks", "INT", "-", "Default: 0"),
            ("overdue_tasks", "INT", "-", "Default: 0"),
            ("last_updated", "TIMESTAMP", "-", "Default: CURRENT_TIMESTAMP"),
        ],
    },
    {
        "num": "4.14", "title": "Aktivitas Tugas", "nama": "task_activities",
        "pk": "id",
        "fk": [
            "task_id (terhubung dengan tasks)",
            "user_id (terhubung dengan users)",
        ],
        "rows": [
            ("id", "SERIAL", "-", "PK"),
            ("task_id", "INT", "-", "FK, NOT NULL"),
            ("user_id", "INT", "-", "FK, NOT NULL"),
            ("message", "TEXT", "-", "NOT NULL"),
            ("activity_type", "VARCHAR", "50", "Default: note"),
            ("file_path", "VARCHAR", "1000", ""),
            ("file_name", "VARCHAR", "500", ""),
            ("file_type", "VARCHAR", "50", ""),
            ("file_size", "INT", "-", "Default: 0"),
            ("created_at", "TIMESTAMP", "-", "Default: CURRENT_TIMESTAMP"),
        ],
    },
    {
        "num": "4.15", "title": "Eskalasi", "nama": "escalations", "pk": "id",
        "fk": [
            "task_id (terhubung dengan tasks)",
            "project_id (terhubung dengan projects)",
            "reported_by (terhubung dengan users)",
            "resolved_by (terhubung dengan users)",
        ],
        "rows": [
            ("id", "SERIAL", "-", "PK"),
            ("task_id", "INT", "-", "FK, NOT NULL"),
            ("project_id", "INT", "-", "FK, NOT NULL"),
            ("reported_by", "INT", "-", "FK, NOT NULL"),
            ("title", "VARCHAR", "500", "NOT NULL"),
            ("description", "TEXT", "-", "NOT NULL"),
            ("status", "VARCHAR", "50", "Default: open"),
            ("priority", "VARCHAR", "50", "Default: medium"),
            ("file_path", "VARCHAR", "1000", ""),
            ("file_name", "VARCHAR", "500", ""),
            ("file_type", "VARCHAR", "50", ""),
            ("file_size", "INT", "-", "Default: 0"),
            ("resolved_by", "INT", "-", "FK"),
            ("resolved_at", "TIMESTAMP", "-", ""),
            ("resolution_notes", "TEXT", "-", ""),
            ("created_at", "TIMESTAMP", "-", "Default: CURRENT_TIMESTAMP"),
            ("updated_at", "TIMESTAMP", "-", "Default: CURRENT_TIMESTAMP"),
        ],
    },
    {
        "num": "4.16", "title": "Audit Log", "nama": "audit_log", "pk": "id",
        "fk": ["changed_by (terhubung dengan users)"],
        "rows": [
            ("id", "SERIAL", "-", "PK"),
            ("entity_type", "VARCHAR", "50", "NOT NULL"),
            ("entity_id", "INT", "-", "NOT NULL"),
            ("entity_name", "VARCHAR", "255", ""),
            ("action", "VARCHAR", "50", "NOT NULL"),
            ("field_name", "VARCHAR", "100", ""),
            ("old_value", "TEXT", "-", ""),
            ("new_value", "TEXT", "-", ""),
            ("changed_by", "INT", "-", "FK, NOT NULL"),
            ("changed_by_name", "VARCHAR", "255", ""),
            ("created_at", "TIMESTAMP", "-", "Default: CURRENT_TIMESTAMP"),
        ],
    },
]

INTRO = (
    "Perancangan model fisik basis data merupakan tahap akhir dalam proses desain "
    "basis data. Pada tahap ini, skema logis atau model Entity Relationship Diagram "
    "(ERD) yang telah dirancang sebelumnya diwujudkan menjadi basis data nyata "
    "menggunakan perangkat lunak manajemen basis data (DBMS) yang dipilih. Tujuan "
    "utama dari perancangan fisik ini adalah untuk mencapai efisiensi dalam "
    "pemrosesan dan pengelolaan data. Berikut adalah perancangan model fisik basis "
    "data untuk sistem manajemen proyek di PT Smart Home Inovasi (SHI)."
)


def set_cell_shading(cell, color_hex: str) -> None:
    tc_pr = cell._tc.get_or_add_tcPr()
    shd = OxmlElement("w:shd")
    shd.set(qn("w:val"), "clear")
    shd.set(qn("w:color"), "auto")
    shd.set(qn("w:fill"), color_hex)
    tc_pr.append(shd)


def set_cell_borders(cell) -> None:
    tc_pr = cell._tc.get_or_add_tcPr()
    tc_borders = OxmlElement("w:tcBorders")
    for side in ("top", "left", "bottom", "right"):
        b = OxmlElement(f"w:{side}")
        b.set(qn("w:val"), "single")
        b.set(qn("w:sz"), "4")
        b.set(qn("w:color"), "000000")
        tc_borders.append(b)
    tc_pr.append(tc_borders)


def style_run(run, *, bold: bool = False, size: int = 11, italic: bool = False) -> None:
    run.font.name = FONT
    run.font.size = Pt(size)
    run.bold = bold
    run.italic = italic
    rpr = run._element.get_or_add_rPr()
    r_fonts = rpr.find(qn("w:rFonts"))
    if r_fonts is None:
        r_fonts = OxmlElement("w:rFonts")
        rpr.append(r_fonts)
    for attr in ("w:ascii", "w:hAnsi", "w:cs", "w:eastAsia"):
        r_fonts.set(qn(attr), FONT)


def write_cell(cell, text: str, *, bold: bool = False, center: bool = False,
               header: bool = False) -> None:
    cell.text = ""
    p = cell.paragraphs[0]
    if center:
        p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    run = p.add_run(text)
    style_run(run, bold=bold or header, size=11)
    cell.vertical_alignment = WD_ALIGN_VERTICAL.CENTER
    set_cell_borders(cell)
    if header:
        set_cell_shading(cell, "D9E2F3")


def add_para(doc, text: str, *, bold: bool = False, italic: bool = False,
             center: bool = False, justify: bool = False, indent_first: float = 0.0,
             indent_left: float = 0.0, size: int = 11, space_before: int = 0,
             space_after: int = 6, heading_level: int | None = None):
    p = doc.add_paragraph()
    if heading_level is not None:
        p.style = doc.styles[f"Heading {heading_level}"]
    if center:
        p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    elif justify:
        p.alignment = WD_ALIGN_PARAGRAPH.JUSTIFY
    pf = p.paragraph_format
    if indent_first:
        pf.first_line_indent = Cm(indent_first)
    if indent_left:
        pf.left_indent = Cm(indent_left)
    pf.space_before = Pt(space_before)
    pf.space_after = Pt(space_after)
    run = p.add_run(text)
    style_run(run, bold=bold, italic=italic, size=size)
    return p


def set_col_widths(table, widths_twips: list[int]) -> None:
    for row in table.rows:
        for idx, cell in enumerate(row.cells):
            cell.width = Twips(widths_twips[idx])


def build_table(doc, t: dict) -> None:
    rows = t["rows"]
    table = doc.add_table(rows=len(rows) + 1, cols=5)
    table.alignment = WD_TABLE_ALIGNMENT.CENTER
    table.autofit = False

    headers = ("No", "Nama Field", "Tipe Data", "Panjang Karakter", "Constraint")
    for i, h in enumerate(headers):
        write_cell(table.rows[0].cells[i], h, header=True, center=True)

    for r_idx, (field, dtype, length, constraint) in enumerate(rows, start=1):
        write_cell(table.rows[r_idx].cells[0], str(r_idx), center=True)
        write_cell(table.rows[r_idx].cells[1], field)
        write_cell(table.rows[r_idx].cells[2], dtype)
        write_cell(table.rows[r_idx].cells[3], length)
        write_cell(table.rows[r_idx].cells[4], constraint, center=True)

    set_col_widths(table, COL_W)


def main() -> None:
    doc = Document()

    style = doc.styles["Normal"]
    style.font.name = FONT
    style.font.size = Pt(12)
    rpr = style.element.get_or_add_rPr()
    r_fonts = rpr.find(qn("w:rFonts"))
    if r_fonts is None:
        r_fonts = OxmlElement("w:rFonts")
        rpr.append(r_fonts)
    for attr in ("w:ascii", "w:hAnsi", "w:cs", "w:eastAsia"):
        r_fonts.set(qn(attr), FONT)

    section = doc.sections[0]
    section.top_margin = Cm(2.54)
    section.bottom_margin = Cm(2.54)
    section.left_margin = Cm(2.54)
    section.right_margin = Cm(2.54)

    add_para(doc, "4.3.3  Perancangan Fisik Basis Data", bold=True, size=12,
             space_before=12, space_after=12)

    add_para(doc, INTRO, justify=True, indent_first=1.27, size=12, space_after=12)

    for idx, t in enumerate(TABLES, start=1):
        add_para(
            doc, f"{idx}.  Tabel {t['title']}", bold=True, size=12,
            indent_left=0.75, space_before=10, space_after=4,
        )
        add_para(doc, f"Nama Tabel: {t['nama']}", indent_left=1.5, size=12,
                 space_after=2)
        add_para(doc, f"Primary Key: {t['pk']}", indent_left=1.5, size=12,
                 space_after=2)
        for fk in t["fk"]:
            add_para(doc, f"Foreign Key: {fk}", indent_left=1.5, size=12,
                     space_after=2)
        add_para(doc, f"Tabel {t['num']} {t['title']}", italic=True, center=True,
                 size=11, space_before=6, space_after=4)
        build_table(doc, t)
        add_para(doc, "", size=11, space_after=0)

    doc.save(OUT)
    print(f"OK {OUT}")


if __name__ == "__main__":
    main()

File: diagram/ai/_gen_fisik_docx.js
const fs = require("fs");
const {
  Document, Packer, Paragraph, TextRun, Table, TableRow, TableCell,
  AlignmentType, HeadingLevel, BorderStyle, WidthType, ShadingType,
  LevelFormat,
} = require("docx");

const border = { style: BorderStyle.SINGLE, size: 4, color: "000000" };
const borders = { top: border, bottom: border, left: border, right: border };

const COL = [600, 2400, 1800, 3360, 1200];
const TBL_W = COL.reduce((a, b) => a + b, 0);

function p(text, opts = {}) {
  return new Paragraph({
    children: [new TextRun({ text, ...opts })],
    spacing: { after: 120 },
    ...(opts.paragraph || {}),
  });
}

function cell(text, opts = {}) {
  const widthIdx = opts.widthIdx ?? 0;
  return new TableCell({
    borders,
    width: { size: COL[widthIdx], type: WidthType.DXA },
    margins: { top: 60, bottom: 60, left: 100, right: 100 },
    shading: opts.header ? { fill: "D9E2F3", type: ShadingType.CLEAR } : undefined,
    verticalAlign: "center",
    children: [new Paragraph({
      alignment: opts.align || AlignmentType.LEFT,
      children: [new TextRun({ text, bold: !!opts.header, size: 22 })],
    })],
  });
}

function headerRow() {
  return new TableRow({
    tableHeader: true,
    children: [
      cell("No", { header: true, align: AlignmentType.CENTER, widthIdx: 0 }),
      cell("Nama Field", { header: true, align: AlignmentType.CENTER, widthIdx: 1 }),
      cell("Tipe Data", { header: true, align: AlignmentType.CENTER, widthIdx: 2 }),
      cell("Panjang Karakter", { header: true, align: AlignmentType.CENTER, widthIdx: 3 }),
      cell("Constraint", { header: true, align: AlignmentType.CENTER, widthIdx: 4 }),
    ],
  });
}

function dataRow(no, field, type, length, constraint) {
  return new TableRow({
    children: [
      cell(String(no), { align: AlignmentType.CENTER, widthIdx: 0 }),
      cell(field, { widthIdx: 1 }),
      cell(type, { widthIdx: 2 }),
      cell(length, { widthIdx: 3 }),
      cell(constraint, { widthIdx: 4 }),
    ],
  });
}

function tableCaption(num, name) {
  return new Paragraph({
    alignment: AlignmentType.CENTER,
    spacing: { before: 120, after: 80 },
    children: [new TextRun({ text: `Tabel ${num} ${name}`, italics: true })],
  });
}

function tableHeader(num, title, pk, fkLines = []) {
  const lines = [
    new Paragraph({
      numbering: { reference: "list", level: 0 },
      spacing: { before: 200, after: 80 },
      children: [new TextRun({ text: `Tabel ${title}`, bold: true })],
    }),
    new Paragraph({
      indent: { left: 720 },
      spacing: { after: 40 },
      children: [new TextRun({ text: `Nama Tabel: ${title}` })],
    }),
    new Paragraph({
      indent: { left: 720 },
      spacing: { after: 40 },
      children: [new TextRun({ text: `Primary Key: ${pk}` })],
    }),
  ];
  for (const fk of fkLines) {
    lines.push(new Paragraph({
      indent: { left: 720 },
      spacing: { after: 40 },
      children: [new TextRun({ text: `Foreign Key: ${fk}` })],
    }));
  }
  return lines;
}

function buildTable(rows) {
  return new Table({
    width: { size: TBL_W, type: WidthType.DXA },
    columnWidths: COL,
    rows: [headerRow(), ...rows.map((r, i) => dataRow(i + 1, ...r))],
  });
}

const tables = [
  {
    num: "4.6", title: "User", pk: "id_user", fk: [],
    rows: [
      ["id_user", "BIGINT", "20", "PK"],
      ["name", "VARCHAR", "255", ""],
      ["email", "VARCHAR", "255", "Unique"],
      ["password_hash", "VARCHAR", "255", ""],
      ["role", "ENUM", "'technician','manager','admin'", ""],
      ["is_active", "TINYINT", "1", "Default: 1"],
      ["created_at", "TIMESTAMP", "-", ""],
    ],
  },
  {
    num: "4.7", title: "Klien", pk: "id_klien", fk: [],
    rows: [
      ["id_klien", "BIGINT", "20", "PK"],
      ["name", "VARCHAR", "255", ""],
      ["address", "VARCHAR", "255", ""],
      ["phone", "VARCHAR", "20", ""],
      ["email", "VARCHAR", "255", ""],
      ["created_at", "TIMESTAMP", "-", ""],
    ],
  },
  {
    num: "4.8", title: "Penugasan Proyek", pk: "(id_proyek, id_user)",
    fk: ["id_proyek (terhubung dengan Proyek)", "id_user (terhubung dengan User)"],
    rows: [
      ["id_proyek", "BIGINT", "20", "PK, FK"],
      ["id_user", "BIGINT", "20", "PK, FK"],
      ["assigned_at", "TIMESTAMP", "-", ""],
    ],
  },
  {
    num: "4.9", title: "Proyek", pk: "id_proyek",
    fk: ["id_klien (terhubung dengan Klien)", "created_by (terhubung dengan User)"],
    rows: [
      ["id_proyek", "BIGINT", "20", "PK"],
      ["name", "VARCHAR", "255", ""],
      ["id_klien", "BIGINT", "20", "FK"],
      ["start_date", "DATE", "-", ""],
      ["end_date", "DATE", "-", ""],
      ["status", "ENUM", "'active','completed','on-hold'", "Default: active"],
      ["phase", "ENUM", "'survey','execution'", "Default: survey"],
      ["project_value", "DECIMAL", "15,2", ""],
      ["survey_approved", "TINYINT", "1", "Default: 0"],
      ["created_by", "BIGINT", "20", "FK"],
      ["created_at", "TIMESTAMP", "-", ""],
    ],
  },
  {
    num: "4.10", title: "Kesehatan Proyek", pk: "id_proyek",
    fk: ["id_proyek (terhubung dengan Proyek)"],
    rows: [
      ["id_proyek", "BIGINT", "20", "PK, FK"],
      ["spi_value", "DECIMAL", "5,2", ""],
      ["status", "ENUM", "'green','amber','red'", ""],
      ["deviation_percent", "DECIMAL", "5,2", ""],
      ["actual_progress", "DECIMAL", "5,2", ""],
      ["planned_progress", "DECIMAL", "5,2", ""],
      ["total_tasks", "INT", "11", ""],
      ["completed_tasks", "INT", "11", ""],
      ["last_updated", "TIMESTAMP", "-", ""],
    ],
  },
  {
    num: "4.11", title: "Tugas", pk: "id_tugas",
    fk: [
      "id_proyek (terhubung dengan Proyek)",
      "assigned_to (terhubung dengan User)",
      "created_by (terhubung dengan User)",
    ],
    rows: [
      ["id_tugas", "BIGINT", "20", "PK"],
      ["id_proyek", "BIGINT", "20", "FK"],
      ["name", "VARCHAR", "255", ""],
      ["assigned_to", "BIGINT", "20", "FK"],
      ["status", "ENUM", "'to_do','working_on_it','done'", "Default: to_do"],
      ["due_date", "DATE", "-", ""],
      ["sort_order", "INT", "11", ""],
      ["created_by", "BIGINT", "20", "FK"],
      ["created_at", "TIMESTAMP", "-", ""],
      ["updated_at", "TIMESTAMP", "-", ""],
    ],
  },
  {
    num: "4.12", title: "Bukti Tugas", pk: "id_bukti",
    fk: ["id_tugas (terhubung dengan Tugas)", "uploaded_by (terhubung dengan User)"],
    rows: [
      ["id_bukti", "BIGINT", "20", "PK"],
      ["id_tugas", "BIGINT", "20", "FK"],
      ["file_path", "VARCHAR", "500", ""],
      ["file_name", "VARCHAR", "255", ""],
      ["file_type", "ENUM", "'image','document','video'", ""],
      ["file_size", "INT", "11", ""],
      ["uploaded_by", "BIGINT", "20", "FK"],
      ["uploaded_at", "TIMESTAMP", "-", ""],
    ],
  },
  {
    num: "4.13", title: "Eskalasi", pk: "id_eskalasi",
    fk: [
      "id_proyek (terhubung dengan Proyek)",
      "id_tugas (terhubung dengan Tugas)",
      "reported_by (terhubung dengan User)",
    ],
    rows: [
      ["id_eskalasi", "BIGINT", "20", "PK"],
      ["id_proyek", "BIGINT", "20", "FK"],
      ["id_tugas", "BIGINT", "20", "FK"],
      ["reported_by", "BIGINT", "20", "FK"],
      ["title", "VARCHAR", "255", ""],
      ["description", "TEXT", "-", ""],
      ["priority", "ENUM", "'low','medium','high'", "Default: medium"],
      ["status", "ENUM", "'open','handled','closed'", "Default: open"],
      ["created_at", "TIMESTAMP", "-", ""],
    ],
  },
];

const children = [
  new Paragraph({
    heading: HeadingLevel.HEADING_2,
    spacing: { before: 240, after: 200 },
    children: [new TextRun({ text: "4.3.3 Perancangan Fisik Basis Data", bold: true })],
  }),
  new Paragraph({
    alignment: AlignmentType.JUSTIFIED,
    spacing: { after: 200 },
    indent: { firstLine: 720 },
    children: [new TextRun({
      text: "Perancangan model fisik basis data merupakan tahap terakhir dalam proses desain basis data. Pada tahap ini, skema logis atau model ERD yang telah dirancang sebelumnya diwujudkan menjadi basis data nyata dengan menggunakan perangkat lunak manajemen basis data (DBMS) yang dipilih. Tujuan utama dari perancangan fisik ini adalah untuk mencapai efisiensi dalam pemrosesan dan pengelolaan data. Berikut adalah perancangan model fisik basis data untuk manajemen proyek di PT SHI.",
    })],
  }),
];

for (const t of tables) {
  children.push(...tableHeader(t.num, t.title, t.pk, t.fk));
  children.push(tableCaption(t.num, t.title));
  children.push(buildTable(t.rows));
}

const doc = new Document({
  styles: {
    default: { document: { run: { font: "Times New Roman", size: 24 } } },
    paragraphStyles: [
      {
        id: "Heading2", name: "Heading 2", basedOn: "Normal", next: "Normal", quickFormat: true,
        run: { size: 28, bold: true, font: "Times New Roman" },
        paragraph: { spacing: { before: 200, after: 200 }, outlineLevel: 1 },
      },
    ],
  },
  numbering: {
    config: [{
      reference: "list",
      levels: [{
        level: 0,
        format: LevelFormat.DECIMAL,
        text: "%1)",
        alignment: AlignmentType.LEFT,
        style: { paragraph: { indent: { left: 720, hanging: 360 } } },
      }],
    }],
  },
  sections: [{
    properties: {
      page: {
        size: { width: 11906, height: 16838 },
        margin: { top: 1440, right: 1440, bottom: 1440, left: 1440 },
      },
    },
    children,
  }],
});

Packer.toBuffer(doc).then((buf) => {
  fs.writeFileSync("D:/__CODING/05-MyProjects/__IRENE/shi-crm/diagram/ai/4.3.3_PERANCANGAN_FISIK.docx", buf);
  console.log("OK");
});
